### PR TITLE
Allow builds to add rules for mainDexClasses list generation

### PR DESCRIPTION
**Pull request by skolberg**
- via new setting key dexMainClassesRules
 - additionally added missing keep rule for 'android.support.multidex.**'

### CODE REVIEW

**Comment by skolberg** [comment]
This allows builds to specify additional keep statements to include classes in the main dex list that are not 'found' by proguard with default ruse.

E.g. classes instantiated by reflection.

**Comment by pfn** [comment]
Is there a use-case for this? have you encountered this need? Mostly curious, it seems so very niche.

The android.support.multidex fix is good, though (although might not be necessary with the `Application` being kept)

**Comment by skolberg** [comment]
Well, 

1. It may be niche, but 
   I am using a library that logs via _slf4j_. And for that I have to have `org.slf4j.impl.StaticLoggerBinder` in the _mainDexList_. 
   b. I am also using _Slick_ (don't ask why) which uses reflection to initialize a database driver.
2. `com.android.tools.fd.**` was already there, and I didn't know if I need it or not.
3. 'android.support.multidex.MultiDexExtractor' might be enough, I'll change that end check if it works.

So, yeah, my needs are special, but is the feature that exotic, that it isn't worth including?

I for my part need it that much, that I would have to fallback to my own fork hence forth.

**Comment by pfn** [comment]
It's fine for the most part, I want to understand the use-case first, though. Try without MultiDexExtractor and see if it works (MultiDexExtractor should still be in the main dex).

Drop the fd.**

Otherwise, it's reasonable enough, I want to understand the use-case for this first, is all.

File: src/rules.scala [hunk 1]
```diff
@@ -533,6 +533,19 @@ object Plugin extends sbt.Plugin with PluginFail {
...
+      "-dontoptimize",
+      "-dontpreverify",
+      "-dontwarn **",
+      "-dontnote **",
+      "-forceprocessing",
+      "-keep public class * extends android.app.backup.BackupAgent { <init>(); }",
+      "-keep public class * extends java.lang.annotation.Annotation { *; }",
+      "-keep class android.support.multidex.** { *; }",
+      "-keep class com.android.tools.fd.** { *; }",
+      "-dontnote com.android.tools.fd.**,android.support.multidex.MultiDexExtractor"
```
Comment (pfn) [inline src/rules.scala]: drop the 2 com.android.tools.fd.** lines

verify whether or not everything is correct without any -keep android.support.multidex at all

**Comment by skolberg** [comment]
I dropeed the `fd.**` and some redundant _-dontnotes_. But `multidex.**` (at least for my build) had to stay. Otherwise I get:
```

05-13 06:57:16.950 I/dalvikvm( 3383): Could not find method android.support.multidex.MultiDex$V19.access$000, referenced from method android.support.multidex.MultiDex.installSecondaryDexes
05-13 06:57:16.950 W/dalvikvm( 3383): VFY: unable to resolve static method 511: Landroid/support/multidex/MultiDex$V19;.access$000 (Ljava/lang/ClassLoader;Ljava/util/List;Ljava/io/File;)V
05-13 06:57:16.950 D/dalvikvm( 3383): VFY: replacing opcode 0x71 at 0x000c
05-13 06:57:17.000 D/dalvikvm( 3383): GC_FOR_ALLOC freed 1104K, 17% free 9612K/11468K, paused 49ms, total 49ms
05-13 06:57:17.000 I/dalvikvm( 3383): Could not find method android.support.multidex.MultiDex$V14.access$100, referenced from method android.support.multidex.MultiDex.installSecondaryDexes
05-13 06:57:17.000 W/dalvikvm( 3383): VFY: unable to resolve static method 510: Landroid/support/multidex/MultiDex$V14;.access$100 (Ljava/lang/ClassLoader;Ljava/util/List;Ljava/io/File;)V
05-13 06:57:17.000 D/dalvikvm( 3383): VFY: replacing opcode 0x71 at 0x0016
05-13 06:57:17.000 I/dalvikvm( 3383): Could not find method android.support.multidex.MultiDex$V4.access$200, referenced from method android.support.multidex.MultiDex.installSecondaryDexes
05-13 06:57:17.000 W/dalvikvm( 3383): VFY: unable to resolve static method 512: Landroid/support/multidex/MultiDex$V4;.access$200 (Ljava/lang/ClassLoader;Ljava/util/List;)V
05-13 06:57:17.000 D/dalvikvm( 3383): VFY: replacing opcode 0x71 at 0x001a
05-13 06:57:17.000 I/MultiDex( 3383): VM with version 1.6.0 does not have multidex support
0

...

05-13 06:57:20.040 E/AndroidRuntime( 3383): java.lang.NoClassDefFoundError: android.support.multidex.MultiDex$V19
05-13 06:57:20.040 E/AndroidRuntime( 3383): 	at android.support.multidex.MultiDex.installSecondaryDexes(MultiDex.java:242)
05-13 06:57:20.040 E/AndroidRuntime( 3383): 	at android.support.multidex.MultiDex.install(MultiDex.java:161)
05-13 06:57:20.040 E/AndroidRuntime( 3383): 	at android.support.multidex.MultiDexApplication.attachBaseContext(MultiDexApplication.java:39)
05-13 06:57:20.040 E/AndroidRuntime( 3383): 	at android.app.Application.attach(Application.java:181)
05-13 06:57:20.040 E/AndroidRuntime( 3383): 	at android.app.Instrumentation.newApplication(Instrumentation.java:991)
05-13 06:57:20.040 E/AndroidRuntime( 3383): 	at android.app.Instrumentation.newApplication(Instrumentation.java:975)
05-13 06:57:20.040 E/AndroidRuntime( 3383): 	at android.app.LoadedApk.makeApplication(LoadedApk.java:511)
0
...

```

**Pull request merged**